### PR TITLE
AppNexus: Forward GPID

### PR DESCRIPTION
**Pull request by SyntaxNode**

### CODE REVIEW

File: adapters/appnexus/appnexus.go [hunk 1]
```diff
@@ -285,6 +286,15 @@ func handleLegacyParams(appnexusExt *openrtb_ext.ExtImpAppnexus) {
...
 }
 
+func validateAppnexusExt(appnexusExt *openrtb_ext.ExtImpAppnexus) error {
+	if appnexusExt.PlacementId == 0 && (appnexusExt.InvCode == "" || appnexusExt.Member == "") {
+		return &errortypes.BadInput{
+			Message: "No placement or member+invcode provided",
+		}
+	}
+	return nil
+}
```
Comment (SyntaxNode) [inline adapters/appnexus/appnexus.go:296]: Moved to co-locate with associated code.

File: adapters/appnexus/appnexus.go [hunk 2]
```diff
@@ -389,18 +391,21 @@ func buildRequestImp(imp *openrtb2.Imp, appnexusExt *openrtb_ext.ExtImpAppnexus,
...
+			PrivateSizes:      ext.Bidder.PrivateSizes,
+			ExtInvCode:        ext.Bidder.ExtInvCode,
+			ExternalImpID:     ext.Bidder.ExternalImpId,
+		},
+		GPID: ext.GPID,
+	}
 
 	var err error
-	imp.Ext, err = json.Marshal(&impExt)
+	imp.Ext, err = json.Marshal(impExt)
```
Comment (SyntaxNode) [inline adapters/appnexus/appnexus.go:408]: Not sure why this was an explicit reference.

File: adapters/appnexus/models.go [hunk 3]
```diff
@@ -4,6 +4,12 @@ import (
 	"encoding/json"
 )
 
+// impExt defines the outgoing data contract.
+type impExt struct {
+	Appnexus impExtAppnexus `json:"appnexus"`
+	GPID     string         `json:"gpid,omitempty"`
+}
```
Comment (SyntaxNode) [inline adapters/appnexus/models.go:11]: Moved above it's child struct `impExtAppnexus`.

File: adapters/appnexus/appnexus.go (hunk 1 above)
Comment (bsardo) [inline adapters/appnexus/appnexus.go:296]: Perhaps this is outside of the scope of this PR given this logic is a rewrite but why do we need this function? Isn't this validation already performed upstream during bidder param validation?
```
"oneOf": [{
    "oneOf": [{
      "required": ["placementId"]
    }, {
      "required": ["placement_id"]
    }]
  }, {
    "oneOf": [{
      "required": ["invCode", "member"]
    }, {
      "required": ["inv_code", "member"]
    }]
  }],
```

Are we keeping this logic just so we can be sure `handleLegacyParams` doesn't perform a copy that results in an invalid request?

Comment (hhhjort) [reply adapters/appnexus/appnexus.go:296]: It is to be consistent with Prebid.JS. It is always difficult to get prebid.JS updated in production, and you can't do A/B testing if the js and server adapter parameters are different.

**Review by hhhjort — APPROVED** [review]
LGTM

**Review by bsardo — APPROVED** [review]
